### PR TITLE
feat(2450): add annotations to chart metadata

### PR DESCRIPTION
**Pull request by kdada**
For #2450. 

Add field `Options` is more friendly to users and developers. Then we can store some values handled by other system in the field.

### CODE REVIEW

**Comment by ibazhitov** [comment]
Please, check #2631.
I've proposed the similar functionality there.

**Comment by technosophos** [comment]
I feel uncomfortable accepting a PR that doesn't provide an implementation in either Helm or Tiller, and has no demonstrated use case for either Helm or Tiller.

**Comment by bacongobbler** [comment]
The only usefulness I can see in adding arbitrary metadata to a chart would be in the context of `helm search`. A chart maintainer could provide a list of keywords which then a chart user could search for. For example, stable/traefik and stable/nginx-ingress could have keywords like "ingress" and "addon".

However, `helm search` already searches for these keywords in the chart description so adding a new `keywords` metadata field wouldn't be useful (and backwards-incompatible). I'd also be concerned about the scope of something so generic as `options`. It doesn't really provide any value to Helm/Tiller as @technosophos mentioned, and I too would be concerned about providing support on a feature that provides no real value to users.

Is there a reason you cannot add this static metadata as hardcoded annotations in your application? Perhaps a more solid use case would be useful to help understand the reason for something like this.

**Comment by kdada** [comment]
Why we need `options` in chart:
Helm says `Helm is a tool for managing Kubernetes charts`. We'll run many tiller instances for our kubernetes clusters. Then we can install/uninstall/update/... applications by helm cmd. It's a fancy tool for users to manage complex commercial applications.

But in many scenarios, we should create our own charts and record some extra info of creators. The extra info may be company, department and so on.
As @bacongobbler said, the field may not very useful for a single user. But it's essential for a group or an automatical platform based on helm.

I can give a demo If we don't have the the field:
I assumes that we have two clusters. One for dev and the other for production. We have two helm registry (as charts repo) in the two clusters. The registry in dev cluster stores all charts for devlopment and it should have a db to store the extra info. Once a chart is stable and need to publish to production cluster, We need create a new chart with an extra file `data.yaml`. When the registry in production cluster received the charts, It should decompress the chart and extract the file to recover the extra info.

If helm can support adding extra key-value data, users and third party systems can use the field to store more infomation. It let helm more flexible.

**Comment by kdada** [comment]
If you still think the field is not useful, I'll close the pr.

**Comment by technosophos** [comment]
Understood. I am also reading your comments on #2631.  Thank you for providing a user story.

/cc @prydonius -- is this helpful for Monocular, too?

**Comment by prydonius** [comment]
> The only usefulness I can see in adding arbitrary metadata to a chart would be in the context of helm search. A chart maintainer could provide a list of keywords which then a chart user could search for. For example, stable/traefik and stable/nginx-ingress could have keywords like "ingress" and "addon". 

Just FYI, The metadata already has a [`keywords`](https://github.com/kubernetes/helm/blob/master/docs/charts.md#the-chartyaml-file) field that takes an array and is used for this purpose.

> is this helpful for Monocular, too?

Given this is for arbitrary metadata, I'm not sure of the best way to represent it in Monocular, I'll review the usecases to see what makes sense.

FWIW, I think that `annotations` is a better name for this field, `options` could sound like a set of options a chart takes.

**Comment by kdada** [comment]
`annotations` is better than `options`. I rename the field to `annotations` and add basic tests for it.

File: pkg/chartutil/chartfile_test.go [hunk 1]
```diff
@@ -88,6 +88,18 @@ func verifyChartfile(t *testing.T, f *chart.Metadata, name string) {
...
+	if len(f.Annotations) != 2 {
+		t.Fatalf("Unexpected annotations")
+	}
+
+	if value := f.Annotations["extrakey"]; value != "extravalue" {
+		t.Errorf("Expected extravalue, but got %s", value)
+	}
+
+	if value := f.Annotations["anotherkey"]; value != "anothervalue" {
+		t.Errorf("Expected anotherkey, but got %s", value)
```
Comment (seh) [inline pkg/chartutil/chartfile_test.go]: - [ ] I think you meant "Expected anothervalue."  
  Instead:
```go
if want, got := "anothervalue", f.Annotations["anotherkey"]; got != want {
  t.Errorf("want %q, got %q", want, got)
}
```

File: _proto/hapi/chart/metadata.proto [hunk 2]
```diff
@@ -80,4 +80,7 @@ message Metadata {
 	// TillerVersion is a SemVer constraints on what version of Tiller is required.
 	// See SemVer ranges here: https://github.com/Masterminds/semver#basic-comparisons
 	string tillerVersion = 15;
+
+	// Annotations is an unstructured key value map for user to store arbitrary data.
```
Comment (seh) [inline _proto/hapi/chart/metadata.proto]: - [ ] Instead:
  > Annotations are additional mappings uninterpreted by Tiller, made available for inspection by other applications.

**Comment by kdada** [comment]
@seh PTAL

File: pkg/chartutil/chartfile_test.go [hunk 3]
```diff
@@ -92,12 +92,12 @@ func verifyChartfile(t *testing.T, f *chart.Metadata, name string) {
 		t.Fatalf("Unexpected annotations")
 	}
 
-	if value := f.Annotations["extrakey"]; value != "extravalue" {
-		t.Errorf("Expected extravalue, but got %s", value)
+	if want, got := "extravalue", f.Annotations["extrakey"]; want != got {
+		t.Errorf("Want %q, but got %q", want, got)
```
Comment (seh) [inline pkg/chartutil/chartfile_test.go]: - [ ] The Go style I'm used to does not capitalize these terse failure messages that are not complete sentences, but if the rest of the Helm project _does_ capitalize them, then this is fine as is.  
  
  I see that elsewhere in this file, the messages are capitalized.

File: pkg/chartutil/chartfile_test.go [hunk 4]
```diff
@@ -92,12 +92,12 @@ func verifyChartfile(t *testing.T, f *chart.Metadata, name string) {
...
 	}
 
-	if value := f.Annotations["anotherkey"]; value != "anothervalue" {
-		t.Errorf("Expected anotherkey, but got %s", value)
+	if want, got := "anothervalue", f.Annotations["anotherkey"]; want != got {
+		t.Errorf("Want %q, but got %q", want, got)
```
Comment (seh) [inline pkg/chartutil/chartfile_test.go]: - [ ] Likewise with the capitalization.

**Review by seh — APPROVED** [review]

**Comment by kdada** [comment]
@technosophos 
Any progress about this pr and #2631?

**Comment by technosophos** [comment]
@seh if you are good with this one, I think it can be merged for 2.7

**Comment by seh** [comment]
I approved the latest round on on 14 July 2017. Consider me satisfied.

**Pull request merged**